src/net_processing.cpp

```diff
@@ -4719,6 +4719,16 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
 
         LogPrint(BCLog::NET, "received block %s peer=%d\n", pblock->GetHash().ToString(), pfrom.GetId());
 
+        const CBlockIndex* prev_block{WITH_LOCK(m_chainman.GetMutex(), return m_chainman.m_blockman.LookupBlockIndex(pblock->hashPrevBlock))};
+
+        if (IsBlockMutated(/*block=*/*pblock,
+                           /*check_witness_root=*/DeploymentActiveAfter(prev_block, m_chainman, Consensus::DEPLOYMENT_SEGWIT))) {
```

Comment:
49257c0304828a185c273fcb99742c54bbef0c8e: `DeploymentActiveAfter` is going to be `false` for a block that's disconnected from anything we have. Assuming the block has witness data, the "No witness data is allowed in blocks that don't commit to witness data" check in `CheckWitnessMalleation` will then get triggered.

So IIUC this could reject a valid block, for the wrong reason. Before this change, the `CheckWitnessMalleation()` was called from `ContextualCheckBlock()`, which in turn is guarded by `prev-blk-not-found`.

I guess you have to repeat the `prev-blk-not-found` (and `bad-prevblk`?) checks here.

Reply:
> So IIUC this could reject a valid block, for the wrong reason.

I assume you mean invalid?

This behavior change would manifest itself if someone is sending a full unconnected(or built on invalid) block unilaterally. And instead of adding 10 DoS points for missing, we outright ban for "mutation"?

Reply:
No, I meant valid but disconnected. Although not normal behavior, you might have the chain up to block N, and some peer could send N + 2 without first sending (headers for) N + 1.

Reply:
I mean, our node calls that invalid :)

opened https://github.com/bitcoin/bitcoin/pull/29524 as what I think a clean fix is, with test coverage